contracts/Marketplace.sol

```diff
@@ -413,4 +417,64 @@ contract Marketplace is Ownable, IMarketplace {
     function reInitialize(address datacoinAddress, address currencyUpdateAgentAddress, address prev_marketplace_address) public onlyOwner {
         _initialize(datacoinAddress, currencyUpdateAgentAddress, prev_marketplace_address);
     }
+
+    //whitelist functionality
+
+    //should the whitelist getter be built into getProduct()?
+    function getRequiresWhitelist(bytes32 productId) public view returns (bool) {
```

Comment:
should I include requiresWhitelist bool in getProduct() instead, which would change the signature?

Reply:
I think that would be better, yes. It's not all that breaking anyway, since the values from events are read by argument names.

Reply:
ok, but see my comment above re createProduct. This would mean that Marketplace1 has a different API than future marketplaces. Not a huge problem IMO.

Reply:
Yea I think now is a good time to learn from the experiences of the first marketplace and start thinking about the IMarketplace from the future-proofing perspective, too. Right now I think it's fine to consider the first marketplace ephemeral (since it's on its way out anyway), but is there something we could do to make this Marketplace more future-proofed, or will createProduct's signature change every time a feature is added to Product? Should we indeed have many createProducts with different signatures?

I think it's maybe not all that bad if each Marketplace has its own ABI, as long as chaining backwards works (and it will because each Marketplace will only "migrate" from the previous and trust it to handle the previous migration(s)).

Of course this would be different if we'd do the upgrading with proxy contracts, in which case we really would swap off implementation, and then we'd have to be more careful to not break compatibility.

Reply:
> I think it's maybe not all that bad if each Marketplace has its own ABI, as long as chaining backwards works (and it will because each Marketplace will only "migrate" from the previous and trust it to handle the previous migration(s)).

I agree with the above. Assuming no other "users" of the contracts than our own dapp, it's enough for a new version to know the interface of the previous one for querying, and products will only ever be created on the newest marketplace.